apps/settings/js/panels/usb_storage/usb_transfer.js

```diff
@@ -33,10 +46,95 @@ define(function(require) {
      * @access public
      * @memberOf UsbTransfer.prototype
      */
-    init: function ut_init() {
+    init: function ut_init(elements, option) {
+      this._elements = elements;
+      if (option && option.usbHotProtocolSwitch !== undefined) {
+        this._usbHotProtocolSwitch = option.usbHotProtocolSwitch;
+      }
       // handle protocol changed function
       SettingsListener.observe(this._keyTransferProtocol, this.PROTOCOL_UMS,
         this._configProtocol.bind(this));
+
+      SettingsListener.observe(this._keyUmsEnabled, false,
+        this._umsEnabledHandler.bind(this));
+
+      this._elements.usbEnabledCheckBox.addEventListener('change',
+        this._umsCheckboxChange.bind(this));
+    },
+
+    _umsCheckboxChange: function storage_umsCheckboxChange(evt) {
+      var checkbox = evt.target;
+      var cset = {};
+      var warningKey = 'ums-turn-on-warning';
+      Debug('state change ' + checkbox.checked);
+      if (checkbox.checked) {
+        // show warning dialog first time
+        AsyncStorage.getItem(warningKey, (showed) => {
+          if (!showed) {
+            Debug('turn-on warning is not showed yet');
+            DialogService.confirm('ums-confirm', {
+              title: 'ums-warning-title',
+              submitButton: 'ok',
+              cancelButton: 'cancel'
+            }).then((result) => {
+              var type = result.type;
+              if (type === 'submit') {
+                Debug('turn on success');
+                AsyncStorage.setItem(warningKey, true);
+                SettingsCache.getSettings(
+                  this._openIncompatibleSettingsDialogIfNeeded.bind(this));
+              } else {
+                Debug('turn on fail');
+                cset[this._keyUmsEnabled] = false;
+                Settings.mozSettings.createLock().set(cset);
+
+                checkbox.checked = false;
+              }
+            });
+          } else {
+            Debug('turn-on warning is showed');
+            SettingsCache.getSettings(
+              this._openIncompatibleSettingsDialogIfNeeded.bind(this));
+          }
+        });
+      } else {
+        cset[this._keyUmsEnabled] = false;
+        Settings.mozSettings.createLock().set(cset);
+      }
+    },
+
+    _openIncompatibleSettingsDialogIfNeeded:
+      function storage_openIncompatibleSettingsDialogIfNeeded(settings) {
+        var cset = {};
+        var umsSettingKey = this._keyUmsEnabled;
+        var usbTetheringSetting = settings['tethering.usb.enabled'];
```

Comment:
Nit: move var oldSetting = 'tethering.usb.enabled'; to the beginning of the function